plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py

```diff
@@ -295,63 +297,64 @@ def remove(self, hash_id: str) -> Dict:
                 f"Error while performing garbage collection: {str(e)}"
             ) from e
 
-    def download(self, hash_id: str, target_dir: str, fix_path: bool = True) -> str:
+    def download(
+        self,
+        hash_id: str,
+        target_dir: Union[str, Path],
+        fix_path: bool = True,
+        attempts: int = 5,
+    ) -> str:
         """
-        Download dir by it's hash.
+        Download dir by its hash.
 
         :param hash_id: str. hash of file to download
-        :param target_dir: str. directory to place downloaded
+        :param target_dir: Union[str, Path]. directory to place downloaded
         :param fix_path: bool. default True. on download don't wrap result in to hash_id directory.
+        :param attempts: int. default 5. How often to attempt the download.
         :return: downloaded path
         """
-        if not os.path.exists(target_dir):  # pragma: nocover
-            os.makedirs(target_dir, exist_ok=True)
 
-        if os.path.exists(os.path.join(target_dir, hash_id)):  # pragma: nocover
-            raise DownloadError(f"{hash_id} was already downloaded to {target_dir}")
-
-        try:
-            self.client.get(hash_id, target_dir)
-        except ipfshttpclient.exceptions.StatusError as e:
-            raise DownloadError(f"error on download {str(e)}") from e
-        downloaded_path = str(Path(target_dir) / hash_id)
-
-        package_path = None
-        if os.path.isdir(downloaded_path):
-            downloaded_files = os.listdir(downloaded_path)
-            if len(downloaded_files) > 0:
-                package_name, *_ = os.listdir(downloaded_path)
-                package_path = str(Path(target_dir) / package_name)
-
-        if package_path is None:
+        target_dir = Path(target_dir)
+        target_dir.mkdir(exist_ok=True, parents=True)
+
+        target_path = target_dir / hash_id
+        if target_path.exists():
+            raise DownloadError(f"{hash_id} was already downloaded to {target_path}")
+
+        while attempts:
+            attempts -= 1
+            try:  # download to tmp_dir in case of midway download failure
+                with tempfile.TemporaryDirectory() as tmp_dir:
+                    self.client.get(hash_id, tmp_dir)
+                    download_path = Path(tmp_dir) / hash_id
+                    if fix_path and (download_path / hash_id).is_dir():
+                        download_path = download_path / hash_id
+                    shutil.copytree(download_path, target_path)
+                    break
+            except ipfshttpclient.exceptions.StatusError as e:
+                logging.error(f"error on download of {hash_id}: {e}")
+        else:
+            raise DownloadError(f"Failed to download: {hash_id}")
+
+        if target_path.is_dir() and any(target_path.iterdir()):
+            package_name, *_ = target_path.iterdir()
+            package_path = Path(target_dir) / package_name
+        else:
             package_path = target_dir
```

Comment:
only God knows what purpose this serves.

https://github.com/valory-xyz/open-aea/blob/06094af29028a3906197c9ac2d3950e39f95dce4/plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py#L327-L335